File: go/stack/stack.go
package stack

type Stack struct {
	dataStore []interface{}
	theSize   int
}

func New() *Stack {
	s := new(Stack)
	s.dataStore = make([]interface{}, 0)
	s.theSize = 0

	return s
}

func (s *Stack) Size() int {
	return s.theSize
}

func (s *Stack) IsEmpty() bool {
	return s.theSize == 0
}

func (s *Stack) Push(element interface{}) {
	s.dataStore = append(s.dataStore, element)
	s.theSize++
}

func (s *Stack) Pop() interface{} {
	if s.IsEmpty() {
		return nil
	}
	last := s.dataStore[s.Size()-1]
	s.dataStore = s.dataStore[:s.Size()-1]
	s.theSize--
	return last
}

func (s *Stack) Peek() interface{} {
	if s.IsEmpty() {
		return nil
	}
	return s.dataStore[s.Size()-1]
}

func (s *Stack) Clear() {
	s.dataStore = s.dataStore[:1]
	s.theSize = 0
}

File: java/stack/Stack.java
package stack;

public class Stack<T> {
	private static final int DEFAULT_CAPACITY = 10;
	
	private T[] dataStore;
	private int top;
	
	public Stack() {
		this.top = 0;
		this.ensureCapacity(DEFAULT_CAPACITY);
	}
	
	public int size() {
		return this.top;
	}
	
	public boolean isEmpty() {
		return this.top == 0;
	}
	
	public void push(T element) {
		if (this.top == this.dataStore.length) {
			this.ensureCapacity(2*this.size());
		}
		this.dataStore[this.top++] = element;
	}
	
	public T pop() {
		if (this.isEmpty())  return null;
		
		return this.dataStore[--this.top];
	}
	
	public T peek() {
		if (this.isEmpty())  return null;
		
		return this.dataStore[this.top-1];
	}
	
	public void clear() {
		this.top = 0;
	}

	@SuppressWarnings("unchecked")
	private void ensureCapacity(int newCapacity) {
		if (newCapacity < this.size()) return;
		
		T[] old = this.dataStore;
		this.dataStore = (T[])new Object[newCapacity];
		for (int i = 0; i < this.size(); i++) {
			this.dataStore[i] = old[i];
		}
	}
}
